src/fips.rs

```diff
@@ -752,12 +752,19 @@ impl ProviderSignatureCtx {
                 Some(f) => {
                     let mut siglen = 0usize;
                     let siglen_ptr: *mut usize = &mut siglen;
-                    res_to_err!(f(
+                    let res = f(
                         self.ctx,
                         signature.as_mut_ptr() as *mut c_uchar,
                         siglen_ptr,
                         signature.len()
-                    ))
+                    );
+                    if res != 1 {
+                        err_rv!(CKR_DEVICE_ERROR);
+                    }
+                    /* update to match the real size -- for example ECDSA signatures may vary in
+                     * length */
+                    signature.resize(siglen, 0);
```

Comment:
unsafe { signature.set_len(siglen) }; ?

Reply:
ah we pass in a slice ... I guess we may have to pass in a vector or separate CK_BYTE_PTR and CK_ULONG_PTR

Reply:
Resolved somehow. But there might be better way.